ys001-yggdrasil-core-specification.md

```diff
@@ -0,0 +1,861 @@
+When responding to a session ping in an attempt to establish a new session, a
+node **must** generate an ephemeral set of `curve25519` session keys, which are
+kept for the lifetime of the session. The ephemeral session public key is then
+sent to the remote side and is used to agree a shared session key, with which
+all session traffic is encrypted.
+
+| Field | Type     | Description                                | Length        |
+|:------|:---------|:-------------------------------------------|:--------------|
+| 1     | `varu64` | Message code: **must** have a value of `5` | 1 byte        |
+| 2     | `bytes`  | Encryption handle                          | 8 bytes       |
+| 3     | `bytes`  | Sender ephemeral session public key        | 32 bytes      |
+| 4     | `varu64` | Timestamp                                  | 1 to 10 bytes |
+| 5     | `coords` | Sender coordinates                         | Variable      |
+| 6     | `varu64` | Sender maximum supported session MTU       | 1 to 2 bytes  |
+
+#### DHT request
+
+A DHT request is sent when a node on the network wishes to find the
+coordinates of a remote node. It **must** be encapsulated within a protocol
+message.
+
+The sending node **must** know at least a partial Node ID to search the DHT for.
+
+| Field | Type     | Description                                 | Length   |
+|:------|:---------|:--------------------------------------------|:---------|
+| 1     | `varu64` | Message code: **must** have a value of `6`  | 1 byte   |
+| 2     | `coords` | Sender coordinates                          | Variable |
+| 3     | `bytes`  | Known bytes of target Node ID to search for | Variable |
+
+#### DHT response
+
+A DHT response is sent in response to a DHT request when a node on the DHT has
+knowledge that helps us to reach the target node. It **must** be encapsulated
+within a protocol message.
+
+A response may contain more than one set of target candidates, therefore the
+fields n<sub>1</sub> and n<sub>2</sub> may be repeated multiple times.
+
+| Field         | Type     | Description                                | Length   |
+|:--------------|:---------|:-------------------------------------------|:---------|
+| 1             | `varu64` | Message code: **must** have a value of `7` | 1 byte   |
+| 2             | `coords` | Sender coordinates                         | 8 bytes  |
+| 3             | `bytes`  | Known bytes of target Node ID searched for | 64 bytes |
+| n<sub>1</sub> | `bytes`  | Target candidate public encryption key     | 32 bytes |
+| n<sub>2</sub> | `coords` | Target candidate coordinates               | Variable |
+
+### Second-level link protocol message formats
+
+#### Switch update
+
+A switch update is a message that contains a fully signed path containing all
+hops from the root node down to a given node.
+
+The root node sends a switch update message to all directly connected peers,
+containing a timestamp and its own signing key. The switch update message
```

Comment:
is timestamp in unix epoch seconds? or maybe milliseconds?
does this mean that nodes need to have synchronized times, or this timestamp is not absolute but relative to node key of node which sends it?

Reply:
Epoch seconds in practice but I that it just needs to be an increasing number so that an update doesn’t get replayed in the future. This section can definitely be made clearer.